### ISSUE TITLE
Decouple Guzzle from GuzzleHttp\Psr7\Response

### ISSUE THREAD

**Issue by jails** [guzzle/guzzle#1844]
PSR-7 define some minimal requirements for HTTP messages interoperability.
The fact Guzzle is coupled to the `Guzzle from GuzzleHttp\Psr7\Response` class break the initial interoperability goal. Indeed it requires an additional adapter to parse responses (like with libraries which do not rely on PSR-7).

Are you guys ok to let Guzzle to return some custom PSR-7 response instances ?
If so I'm ready to get the job done with some guidance on where/how would you like to have this custom option to be setted.

**Comment by mtdowling** [comment]
How is Guzzle coupled to a specific PSR-7 response, and what's your suggestion?

**Comment by jails** [comment]
Hi, thank you for your interest.
Right now, response instances are created [here](https://github.com/guzzle/guzzle/blob/56bf6c1cc3e1b24691f10a09321516d4790cecbd/src/Handler/EasyHandle.php#L76) and [there](https://github.com/guzzle/guzzle/blob/56bf6c1cc3e1b24691f10a09321516d4790cecbd/src/Handler/StreamHandler.php#L114)

So it would be interesting to be able to choose what kind of response instance we need.

I'm new to Guzzle but could be something like that:

```php
$client = new \GuzzleHttp\Client(['response_factory' => function($status, $headers, $body, $version, $reason) {
    return new MyResponse($body)
        ->withStatus($status, $reason)
        ->setHeaders($headers)
        ->setVersion($version);
}]);
```

**Comment by mtdowling** [comment]
Why?

Guzzle already has a dependency on the guzzlehttp/psr7 package, so I don't think we should introduce the complexity of allowing the creation of custom PSR-7 response instances (though some kind of factory or whatever).

Guzzle is already decoupled from `GuzzleHttp\Psr7\Response` in that the interfaces and contracts Guzzle provides is that it creates PSR-7 responses. It is not tied to `GuzzleHttp\Psr7\Response` and middlewares, for example, are free to create whatever implementation they wish. The fact that `GuzzleHttp\Psr7\Response` responses are created within Guzzle is an implementation detail that doesn't affect consumers.

**Comment by jails** [comment]
The reason would probably be easier to get with a use case in mind:
So I'm working with a [PSR-7 compatible library](https://github.com/crysalead/net) which has some custom features like content negotiation.
Long story short, when I call the `->get()` method on a response it returns some correctly decoded data (json, csv or whatever).

So when I'm using Guzzle as transport layer, I need to perform the following on received responses:

```php
$client = new \GuzzleHttp\Client();
$request = new \Lead\Net\Http\Request::parseUrl('https://foo.com/api//index.php');
$response = $client->send($request);

$myResponse = new \Lead\Net\Http\Request()
        ->withStatus($response->getStatusCode(), $response->getReasonPhrase())
        ->withHeader($response->getHeaders())
        ->withBody($response->getBody())
        ->withProtocolVersion($response->getProtocolVersion());

var_dump($myResponse->get());
```
In short, I need to convert a PSR-7 compatible response into another PSR-7 compatible response which seems to be what PSR-7 attempted to solve.

Well it's not really a big deal since I can wrap all this boilerplate code inside a custom client class by extending `GuzzleHttp\Client`. But I thought it can be solved right at source.

**Comment by sagikazarmark** [comment]
> So I'm working with a PSR-7 compatible library which has some custom features like content negotiation.

I think you are closing from the wrong direction to what interoperability means in this case (although admittedly there is room for improvement).

Guzzle using it's own PSR-7 package keeps the contract of the PSR-7 interfaces and returns objects implementing those them so you can do whatever you want within the boundaries of the standard. That's the only interoperability PSR-7 currently provides.

What you are trying to achieve is not unique and I don't say it's invalid, all I say currently Guzzle does what it's supposed to do.

There are projects for supporting your use case though and eventually we might add support for those in Guzzle too:

- [PSR-15](https://github.com/http-interop/http-factory) interfaces supporting exactly what you are trying to achieve (still a proposal)
- [HTTPlug](https://github.com/php-http/httplug) PSR-7 HTTP client interfaces (PSR draft is WIP)

For now I think you have to wrap the returned returned response (which you can do easily, because it's compatible with PSR-7). For that I suggest you taking a look at this project which provides you helper traits for message decorators:

https://github.com/php-http/message/tree/master/src/Decorator

**Comment by jails** [comment]
Thanks @sagikazarmark for sharing your insights. Interfaces let you separate the definition of objects from their implementation. So to take full advantage of them, implementations has to be injected somehow. Guzzle already allow to inject the PSR-7 request through the `send()` method so I don't think adding another layer of abstraction should be needed to solve a dependency injection issue imo.

**Comment by sagikazarmark** [comment]
> So to take full advantage of them, implementations has to be injected somehow.

I don't fully agree with this statement, at least not with the reasoning behind and I think there is a little bit of contradiction as well. You expect Guzzle to return a specific *implementation*, not an object implementing an interface (which it actually does). As soon as you expect something which is not part of the interface you kind of misuse it IMHO.

Anyway, if you need a solution now, you can write a middleware which wraps the returned response (which is a ResponseInterface implementation, whether it's Guzzle PSR-7 or not should not matter at all) and returns yours.

**Comment by jails** [comment]
And what about simply allowing Guzzle to return raw responses instead ?

```php
$client = new GuzzleHttp\Client(['psr7' => false]);
$res = $client->request('GET', 'http://httpbin.org');
var_dump($res)
// [
//     "status" => ...
//     "headers" => ...
//     "body" => ...
//     "version" => ...
//     "reason" => ...
//  ]
```

**Comment by sagikazarmark** [comment]
That would complicate a lot of things IMHO, so I would rather not go that way.

There is a chance that we will support PSR-15 eventually, till then there are good enough solutions by extending (middleware) or wrapping Guzzle which are far better than returning mixed typed values.

**Comment by jails** [comment]
Thanks @sagikazarmark for your suggestions and your time :+1:
